File: 658_find_K_closest_elements.java
class Solution {
    // Time Complexity: O(log(n) + k)
    // Space Complexity: O(k)   (for generate k eles sub list)
    public List<Integer> findClosestElements(int[] src, int k, int x) {
        int idx = findClosestIndex(src, x);

        List<Integer> arr = new ArrayList<>();
        for (int val: src) arr.add(val);

        if (idx < 0) {
            return arr.subList(0, k);
        } else if (idx >= src.length) {
            return arr.subList(src.length - k, src.length);
        } else {
            // initialize maximum low index (not smaller than 0)
            // and minimum high index (not larger than the length of this array)
            int low = Math.max(0, idx - k - 1), high = Math.min(src.length - 1, idx + k + 1);
            while (high - low + 1 > k) {
                // 程式簡化：考慮 range 的變化，共兩大種類，其中可細分出四種case
                // shrink high if left elements are closer in difference
                if (low < 0 || (x - src[low] <= src[high] - x)) {
                    // if low >= 0 -> smaller eles are available -> high--
                    high--;
                } else if (high > src.length - 1 || (x - src[low] > src[high] - x)) {
                    // if smaller elements are not available
                    // also if high < src.length -> use large elements
                    low++;
                }
            }
            return arr.subList(low, high + 1);
        }
    }

    public int findClosestIndex(int[] arr, int x) {
        int lo = 0, hi = arr.length - 1;
        while (lo < hi) {
            int mid = lo + (hi - lo) / 2;
            if (arr[mid] > x) {
                hi = mid - 1;
            } else if (arr[mid] < x) {
                lo = mid + 1;
            } else {
                return mid;
            }
        }
        return lo;
    }
}

File: python/658_find_k_closest_elements.py
class Solution:
    # Time Complexity: O(N)
    # Space Complexity: O(N)
    def findClosestElements(self, arr: List[int], k: int, x: int) -> List[int]:
        if x < arr[0]:
            return arr[:k]
        elif x >= arr[-1]:
            return arr[len(arr) - k:]
        else:
            curr = 0
            while curr < len(arr):
                if arr[curr] > x:
                    arr.insert(curr, x)
                    break
                curr += 1
            left = curr - 1
            right = curr + 1
            smaller = []
            larger = []
            while left >= 0 and right < len(arr) and k > 0:
                if abs(x - arr[left]) <= abs(x - arr[right]):
                    smaller.insert(0, arr[left])
                    left -= 1
                else:
                    larger.append(arr[right])
                    right += 1
                k -= 1
            if k > 0:
                if left < 0 and right < len(arr):
                    larger += arr[right:right + k]
                elif left >= 0 and right >= len(arr):
                    smaller = arr[left - k + 1:left + 1] + smaller
            return smaller + larger

